// @consensys/linea-sdk/dist/lib/postman/migrations/1687890326970-AddNewColumns.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddNewColumns1687890326970 = void 0;
const typeorm_1 = require("typeorm");
class AddNewColumns1687890326970 {
    async up(queryRunner) {
        await queryRunner.addColumns("message", [
            new typeorm_1.TableColumn({
                name: "claim_number_of_retry",
                type: "integer",
                isNullable: false,
                default: 0,
            }),
            new typeorm_1.TableColumn({
                name: "claim_last_retried_at",
                type: "timestamp with time zone",
                isNullable: true,
            }),
            new typeorm_1.TableColumn({
                name: "claim_gas_estimation_threshold",
                type: "numeric",
                isNullable: true,
            }),
        ]);
    }
    async down(queryRunner) {
        await queryRunner.dropColumn("message", "claim_number_of_retry");
        await queryRunner.dropColumn("message", "claim_last_retried_at");
        await queryRunner.dropColumn("message", "claim_gas_estimation_threshold");
    }
}
exports.AddNewColumns1687890326970 = AddNewColumns1687890326970;
