// @consensys/linea-sdk/dist/lib/postman/utils/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageStatus = exports.Direction = exports.DatabaseRepoName = exports.DatabaseErrorType = void 0;
var DatabaseErrorType;
(function (DatabaseErrorType) {
    DatabaseErrorType["Read"] = "read";
    DatabaseErrorType["Insert"] = "insert";
    DatabaseErrorType["Update"] = "update";
    DatabaseErrorType["Delete"] = "delete";
})(DatabaseErrorType || (exports.DatabaseErrorType = DatabaseErrorType = {}));
var DatabaseRepoName;
(function (DatabaseRepoName) {
    DatabaseRepoName["MessageRepository"] = "MessageRepository";
})(DatabaseRepoName || (exports.DatabaseRepoName = DatabaseRepoName = {}));
var Direction;
(function (Direction) {
    Direction["L1_TO_L2"] = "L1_TO_L2";
    Direction["L2_TO_L1"] = "L2_TO_L1";
})(Direction || (exports.Direction = Direction = {}));
var MessageStatus;
(function (MessageStatus) {
    MessageStatus["SENT"] = "SENT";
    MessageStatus["ANCHORED"] = "ANCHORED";
    MessageStatus["PENDING"] = "PENDING";
    MessageStatus["CLAIMED_SUCCESS"] = "CLAIMED_SUCCESS";
    MessageStatus["CLAIMED_REVERTED"] = "CLAIMED_REVERTED";
    MessageStatus["NON_EXECUTABLE"] = "NON_EXECUTABLE";
    MessageStatus["ZERO_FEE"] = "ZERO_FEE";
    MessageStatus["FEE_UNDERPRICED"] = "FEE_UNDERPRICED";
    MessageStatus["EXCLUDED"] = "EXCLUDED";
})(MessageStatus || (exports.MessageStatus = MessageStatus = {}));
