// @consensys/linea-sdk/dist/lib/postman/utils/mappers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapMessageEntityToMessage = exports.mapMessageToMessageEntity = void 0;
const mapMessageToMessageEntity = (message) => {
    return {
        id: message?.id,
        ...message,
        createdAt: new Date(),
        updatedAt: new Date(),
    };
};
exports.mapMessageToMessageEntity = mapMessageToMessageEntity;
const mapMessageEntityToMessage = (entity) => {
    return {
        messageSender: entity.messageSender,
        destination: entity.destination,
        fee: entity.fee,
        value: entity.value,
        messageNonce: entity.messageNonce,
        calldata: entity.calldata,
        messageHash: entity.messageHash,
        messageContractAddress: entity.messageContractAddress,
        sentBlockNumber: entity.sentBlockNumber,
        direction: entity.direction,
        status: entity.status,
        claimTxCreationDate: entity.claimTxCreationDate,
        claimTxGasLimit: entity.claimTxGasLimit,
        claimTxMaxFeePerGas: entity.claimTxMaxFeePerGas,
        claimTxMaxPriorityFeePerGas: entity.claimTxMaxPriorityFeePerGas,
        claimTxNonce: entity.claimTxNonce,
        claimTxHash: entity.claimTxHash,
        claimNumberOfRetry: entity.claimNumberOfRetry,
        claimLastRetriedAt: entity.claimLastRetriedAt,
        claimGasEstimationThreshold: entity.claimGasEstimationThreshold,
        createdAt: entity.createdAt,
        updatedAt: entity.updatedAt,
    };
};
exports.mapMessageEntityToMessage = mapMessageEntityToMessage;
