// @consensys/linea-sdk/dist/lib/postman/dataSource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DB = void 0;
const typeorm_1 = require("typeorm");
const typeorm_naming_strategies_1 = require("typeorm-naming-strategies");
const Message_entity_1 = require("./entity/Message.entity");
const _1685985945638_InitialDatabaseSetup_1 = require("./migrations/1685985945638-InitialDatabaseSetup");
const _1687890326970_AddNewColumns_1 = require("./migrations/1687890326970-AddNewColumns");
const _1687890694496_UpdateStatusColumn_1 = require("./migrations/1687890694496-UpdateStatusColumn");
const _1689084924789_RemoveUniqueConstraint_1 = require("./migrations/1689084924789-RemoveUniqueConstraint");
const _1701265652528_AddNewIndexes_1 = require("./migrations/1701265652528-AddNewIndexes");
class DB {
    static create(config) {
        return new typeorm_1.DataSource({
            ...config,
            entities: [Message_entity_1.MessageEntity],
            namingStrategy: new typeorm_naming_strategies_1.SnakeNamingStrategy(),
            migrations: [
                _1685985945638_InitialDatabaseSetup_1.InitialDatabaseSetup1685985945638,
                _1687890326970_AddNewColumns_1.AddNewColumns1687890326970,
                _1687890694496_UpdateStatusColumn_1.UpdateStatusColumn1687890694496,
                _1689084924789_RemoveUniqueConstraint_1.RemoveUniqueConstraint1689084924789,
                _1701265652528_AddNewIndexes_1.AddNewIndexes1701265652528,
            ],
            migrationsTableName: "migrations",
            logging: ["error"],
            migrationsRun: true,
        });
    }
}
exports.DB = DB;
