// @consensys/linea-sdk/dist/lib/contracts/EventParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventParser = void 0;
class EventParser {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static parseEvent(event) {
        return {
            args: event.args,
            blockNumber: event.blockNumber,
            logIndex: event.index,
            contractAddress: event.address,
            transactionHash: event.transactionHash,
        };
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static filterAndParseEvents(events) {
        return events.filter((e) => e.removed === false).map(this.parseEvent);
    }
    static parsedEventToMessage(event, direction, status) {
        return {
            messageSender: event.args?._from,
            destination: event.args?._to,
            fee: BigInt(event.args?._fee).toString(),
            value: BigInt(event.args?._value).toString(),
            messageNonce: parseInt(event.args?._nonce.toString()),
            calldata: event.args?._calldata,
            messageHash: event.args?._messageHash,
            messageContractAddress: event.contractAddress,
            sentBlockNumber: event.blockNumber,
            logIndex: event.logIndex,
            claimNumberOfRetry: 0,
            direction,
            status,
        };
    }
}
exports.EventParser = EventParser;
