// @consensys/linea-sdk/dist/lib/postman/transactions/L2ClaimStatusWatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.L2ClaimStatusWatcher = void 0;
const logger_1 = require("../../logger");
const enums_1 = require("../utils/enums");
const ClaimStatusWatcher_1 = require("./ClaimStatusWatcher");
class L2ClaimStatusWatcher extends ClaimStatusWatcher_1.ClaimStatusWatcher {
    constructor(dataSource, l2MessageServiceContract, config, loggerOptions) {
        super(dataSource, l2MessageServiceContract, config, enums_1.Direction.L1_TO_L2);
        this.l2MessageServiceContract = l2MessageServiceContract;
        this.logger = (0, logger_1.getLogger)(L2ClaimStatusWatcher.name, loggerOptions);
    }
    async isRateLimitExceededError(transactionHash) {
        try {
            const tx = await this.l2MessageServiceContract.provider.getTransaction(transactionHash);
            const errorEncodedData = await this.l2MessageServiceContract.provider.call({
                to: tx?.to,
                from: tx?.from,
                nonce: tx?.nonce,
                gasLimit: tx?.gasLimit,
                data: tx?.data,
                value: tx?.value,
                chainId: tx?.chainId,
                accessList: tx?.accessList,
                maxPriorityFeePerGas: tx?.maxPriorityFeePerGas,
                maxFeePerGas: tx?.maxFeePerGas,
            });
            const error = this.l2MessageServiceContract.contract.interface.parseError(errorEncodedData);
            return error?.name === "RateLimitExceeded";
        }
        catch (e) {
            return false;
        }
    }
}
exports.L2ClaimStatusWatcher = L2ClaimStatusWatcher;
