// @consensys/linea-sdk/dist/lib/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.L2_MERKLE_TREE_ADDED_EVENT_SIGNATURE = exports.L2_MESSAGING_BLOCK_ANCHORED_EVENT_SIGNATURE = exports.MESSAGE_SENT_EVENT_SIGNATURE = exports.MESSAGE_CLAIMED_STATUS = exports.MESSAGE_ANCHORED_STATUS = exports.MESSAGE_UNKNOWN_STATUS = exports.DEFAULT_L2_MESSAGE_TREE_DEPTH = exports.PROFIT_MARGIN_MULTIPLIER = exports.DEFAULT_MAX_TX_RETRIES = exports.DEFAULT_MAX_CLAIM_GAS_LIMIT = exports.DEFAULT_RATE_LIMIT_MARGIN = exports.DEFAULT_CALLDATA_ENABLED = exports.DEFAULT_EOA_ENABLED = exports.DEFAULT_RETRY_DELAY_IN_SECONDS = exports.DEFAULT_MAX_NUMBER_OF_RETRIES = exports.DEFAULT_PROFIT_MARGIN = exports.DEFAULT_LISTENER_BLOCK_CONFIRMATIONS = exports.DEFAULT_GAS_LIMIT = exports.DEFAULT_GAS_ESTIMATION_PERCENTILE = exports.DEFAULT_ENFORCE_MAX_GAS_FEE = exports.DEFAULT_MAX_FEE_PER_GAS = exports.DEFAULT_MAX_NONCE_DIFF = exports.DEFAULT_MAX_FETCH_MESSAGES_FROM_DB = exports.DEFAULT_MAX_BLOCKS_TO_FETCH_LOGS = exports.DEFAULT_DB_DAYS_BEFORE_NOW_TO_DELETE = exports.DEFAULT_DB_CLEANING_INTERVAL = exports.DEFAULT_DB_CLEANER_ENABLED = exports.DEFAULT_LISTENER_INTERVAL = exports.DEFAULT_MESSAGE_SUBMISSION_TIMEOUT = void 0;
exports.DEFAULT_MESSAGE_SUBMISSION_TIMEOUT = 300000;
exports.DEFAULT_LISTENER_INTERVAL = 4000;
exports.DEFAULT_DB_CLEANER_ENABLED = false;
exports.DEFAULT_DB_CLEANING_INTERVAL = 43200000;
exports.DEFAULT_DB_DAYS_BEFORE_NOW_TO_DELETE = 14;
exports.DEFAULT_MAX_BLOCKS_TO_FETCH_LOGS = 1000;
exports.DEFAULT_MAX_FETCH_MESSAGES_FROM_DB = 1000;
exports.DEFAULT_MAX_NONCE_DIFF = 10000;
exports.DEFAULT_MAX_FEE_PER_GAS = 100000000000;
exports.DEFAULT_ENFORCE_MAX_GAS_FEE = false;
exports.DEFAULT_GAS_ESTIMATION_PERCENTILE = 20;
exports.DEFAULT_GAS_LIMIT = 0;
exports.DEFAULT_LISTENER_BLOCK_CONFIRMATIONS = 4;
exports.DEFAULT_PROFIT_MARGIN = 1.0;
exports.DEFAULT_MAX_NUMBER_OF_RETRIES = 100;
exports.DEFAULT_RETRY_DELAY_IN_SECONDS = 30;
exports.DEFAULT_EOA_ENABLED = false;
exports.DEFAULT_CALLDATA_ENABLED = false;
exports.DEFAULT_RATE_LIMIT_MARGIN = 0.95;
exports.DEFAULT_MAX_CLAIM_GAS_LIMIT = 100000;
exports.DEFAULT_MAX_TX_RETRIES = 20;
exports.PROFIT_MARGIN_MULTIPLIER = 100;
exports.DEFAULT_L2_MESSAGE_TREE_DEPTH = 5;
exports.MESSAGE_UNKNOWN_STATUS = 0;
exports.MESSAGE_ANCHORED_STATUS = 1;
exports.MESSAGE_CLAIMED_STATUS = 2;
exports.MESSAGE_SENT_EVENT_SIGNATURE = "0xe856c2b8bd4eb0027ce32eeaf595c21b0b6b4644b326e5b7bd80a1cf8db72e6c";
exports.L2_MESSAGING_BLOCK_ANCHORED_EVENT_SIGNATURE = "0x3c116827db9db3a30c1a25db8b0ee4bab9d2b223560209cfd839601b621c726d";
exports.L2_MERKLE_TREE_ADDED_EVENT_SIGNATURE = "0x300e6f978eee6a4b0bba78dd8400dc64fd5652dbfc868a2258e16d0977be222b";
